011-kafka-canary.md

```diff
@@ -0,0 +1,40 @@
+# Strmixi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and continously producing and consuming events on the topics. 
+
+## Current situation
+
+Currently  Strimzi does not have any canary style feature. It does have readiness checks for Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following:
+  
+  * Improved readiness check for Kafka clusters
+  * SLI(s) for service admins
+  * Metrics:
+    * End-to-end latency
+    * Succesfull/failed events
+    * Disk usage
+    * Disk errors
+
+## Proposal
+
+The canary would be implemented as part of the Kafka Roller component. 
+A single _canary_ topic would be created on creation of a Kafka cluster. The topic would have one partition for Kafka broker. Events would then be continously produced and consumed to/from the canary topic. The internal TLS 9091 port would be used to connect to the brokers.
```

Comment:
Mentioning one partition for Kafka broker, what will be the replication factor (RF)? Maybe it would be just RF = 1?
I see a problem when a Kafka broker goes down for any reason (real actual problem or rolling update).
If, let's say broker A, is hosting partition X, the canary tool should detect that that broker A is down. If RF > 1, a leader for that partition will be elected on another broker, let's say broker B ... but it's not much of interest for us, because broker B is already hosting partition Y for canary test. When broker A comes back we want to have it as leader again for partition X (preferred leader election). Does it really make sense having a RF > 1 for this feature then?

Reply:
In out-of-band conversation with Rob I suggested we could use a single topic with RF=1 and #partitions = #brokers × #logdirs-per-broker, i.e. with a partition per log dir. This would mean that leadership change would not happen and (outside of rolling restart) would tell us immediately whether there was a problem writing to or reading from the log dir volume. 

The topic would be tuned to have a short retention and small segment size, so it wouldn't require much in the way of disk space.

On reflection this would cause problems when downscaling, since Kafka doesn't support decreasing the number of partitions. This isn't necessarily a problem. We could just ignore the spare partitions, or we could delete the topic and recreate it, since the actual data in the topic isn't of much value, it's the action of producing and consuming it which is useful.

While it would be configurable, I do think for the purposes of the proposal we should be more concrete about the frequency of message production. Are we talking about a 1s, 5s, 10s or 30s period? I suspect 1s or 5s is reasonable.

Reply:
> In out-of-band conversation with Rob I suggested we could use a single topic with RF=1 and #partitions = #brokers × #logdirs-per-broker, i.e. with a partition per log dir. This would mean that leadership change would not happen and (outside of rolling restart) would tell us immediately whether there was a problem writing to or reading from the log dir volume.

That's exactly what I was explaining with my example ... maybe badly :-)

> ... or we could delete the topic and recreate it ...

Maybe we should pay attention to this because if auto topic creation is enabled (in the broker config) and we delete the topic, because the canary tool is running, we could hit the time period when it's going to produce and the topic is created with the default broker configuration instead of what we need.

Reply:
> In out-of-band conversation with Rob I suggested we could use a single topic with RF=1 and #partitions = #brokers × #logdirs-per-broker, i.e. with a partition per log dir.

Why not use single partition and RF=<Number of brokers> and min.isr=<RF-1>? That actually tests the rolling update works without disruptions and can measure availability. How would you measure availability if the topics are coming and going during rolling updates etc.?

Reply:
That's not worthless, but it's measuring a different thing. It doesn't really tell you that there's something wrong with _this_ particular broker, which is what KafkaRoller is interested in.

Reply:
That is good point ... so you probably need both to achieve what this proposal promises - one for improved broker health information and one for SLIs.